File: src/data/types/purchaseScannerResult.ts
import type { Scanner } from "./scanner.ts";
import type { ScannerPool } from "./scannerPool.ts";
import type { Wallet } from "./wallet.ts";

// Necessary completion, same precedent as PurchaseShipResult: the Scanner
// GDD amendment names `purchaseScanner(scannerId, playerId): Scanner |
// PurchaseError` but never defines PurchaseError, and a bare Scanner
// return on success can't also carry the updated pool/wallet a pure
// function must return alongside it.
export interface PurchaseScannerSucceeded {
  purchased: true;
  scanner: Scanner;
  updatedPool: ScannerPool;
  updatedWallet: Wallet;
}

export interface PurchaseScannerRejected {
  purchased: false;
  reason: string;
}

export type PurchaseScannerResult = PurchaseScannerSucceeded | PurchaseScannerRejected;

File: unity/ProfitableCore/Schema/PurchaseScannerResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/purchaseScannerResult.ts's PurchaseScannerResult =
// PurchaseScannerSucceeded | PurchaseScannerRejected discriminated union.
public abstract class PurchaseScannerResult
{
    public abstract bool Purchased { get; }
}

public sealed class PurchaseScannerSucceeded : PurchaseScannerResult
{
    public override bool Purchased => true;
    public Scanner Scanner { get; init; } = new();
    public ScannerPool UpdatedPool { get; init; } = new();
    public Wallet UpdatedWallet { get; init; } = new();
}

public sealed class PurchaseScannerRejected : PurchaseScannerResult
{
    public override bool Purchased => false;
    public string Reason { get; init; } = string.Empty;
}
